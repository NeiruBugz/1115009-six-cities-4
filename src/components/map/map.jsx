import React, {PureComponent, createRef} from 'react';
import PropTypes from 'prop-types';
import {icon, map, tileLayer, marker} from 'leaflet';
import {connect} from 'react-redux';
import {offerShape} from '../../prop-types/offer.types';

class Map extends PureComponent {
  constructor(props) {
    super(props);
    this.mapRef = createRef();
    this.map = null;
    this._icon = icon({
      iconUrl: `img/pin.svg`,
      iconSize: [30, 30]
    });
    this._activeIcon = icon({
      iconUrl: `img/pin-active.svg`,
      iconSize: [30, 30],
    });
    this._markers = [];
  }

  _createMarkers() {
    const {coordinates, activeCard} = this.props;

    coordinates.forEach((coords) => {
      const mrk = marker(coords, {
        icon: JSON.stringify(coords) === JSON.stringify(activeCard.coordinates) ? this._activeIcon : this._icon
      });
      mrk.addTo(this.map);
      this._markers.push(marker);
    });
  }

  _resetMarkers() {
    this._markers = [];
  }

  componentDidMount() {
    if (!this.mapRef.current) {
      return;
    }

    const {city, zoom} = this.props;

    this.map = map(this.mapRef.current, {
      zoom,
      city,
      zoomControl: false,
      marker: true,
    });

    this.map.setView(city, zoom);

    tileLayer(`https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png`, {
      attribution: `&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors &copy; <a href="https://carto.com/attributions">CARTO</a>`
    }).addTo(this.map);

    this._createMarkers();
  }

  componentDidUpdate() {
    this._resetMarkers();
    this._createMarkers();
  }

  render() {
    return <div ref={this.mapRef} style={{height: `100%`}} />;
  }
}

Map.propTypes = {
  city: PropTypes.arrayOf(PropTypes.number).isRequired,
  zoom: PropTypes.number.isRequired,
  coordinates: PropTypes.arrayOf(PropTypes.arrayOf(PropTypes.number)).isRequired,
  activeCard: PropTypes.shape({}),
};

Map.defaultProps = {
  activeCard: {},
};

const mapStateToProps = (state) => ({
  activeCard: state.activeCard,
});

export {Map};

export default connect(mapStateToProps, null)(Map);
