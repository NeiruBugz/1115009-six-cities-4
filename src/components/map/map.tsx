import React, {PureComponent, createRef, RefObject} from 'react';
import {
  icon,
  map,
  tileLayer,
  marker,
  Map as LeafletMap,
  Icon,
  Marker,
  LatLngTuple,
  LatLngExpression
} from 'leaflet';
import {connect} from 'react-redux';

type MapProps = {
  coordinates: LatLngTuple[] | number[][];
  activeCard: number[];
  city: LatLngTuple | LatLngExpression;
  zoom: number;
};

class Map extends PureComponent<MapProps> {
  mapRef: RefObject<HTMLDivElement>;
  map: LeafletMap;
  _icon: Icon;
  _activeIcon: Icon;
  _markers: Marker[];

  constructor(props) {
    super(props);
    this.mapRef = createRef();
    this.map = null;
    this._icon = icon({
      iconUrl: `img/pin.svg`,
      iconSize: [27, 39]
    });
    this._activeIcon = icon({
      iconUrl: `img/pin-active.svg`,
      iconSize: [27, 39],
    });
    this._markers = [];
  }

  _createMarkers() {
    const {coordinates, activeCard} = this.props;

    coordinates.forEach((coords) => {
      const mrk = marker(coords, {
        icon: JSON.stringify(coords) === JSON.stringify(activeCard) ? this._activeIcon : this._icon
      });
      mrk.addTo(this.map);
      // @ts-ignore
      this._markers.push(marker);
    });
  }

  _resetMarkers() {
    this._markers = [];
  }

  componentDidMount() {
    if (!this.mapRef.current) {
      return;
    }

    const {city, zoom} = this.props;

    this.map = map(this.mapRef.current, {
      zoom,
      center: city,
      zoomControl: false,
    });

    this.map.setView(city, zoom);

    tileLayer(`https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png`, {
      attribution: `&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors &copy; <a href="https://carto.com/attributions">CARTO</a>`
    }).addTo(this.map);

    this._createMarkers();
  }

  componentDidUpdate() {
    this._resetMarkers();
    this._createMarkers();
  }

  render() {
    return <div ref={this.mapRef} style={{height: `100%`}} />;
  }
}

const mapStateToProps = (state) => ({
  activeCard: state.activeCard,
});

export {Map};

export default connect(mapStateToProps, null)(Map);
