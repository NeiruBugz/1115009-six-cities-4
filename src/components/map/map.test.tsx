import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { Map } from './map';
import { LatLngTuple } from "leaflet";

const mockMapProps = {
  city: [52.38333, 4.9],
  zoom: 12,
  coordinates: [
    [52.3909553943508, 4.85309666406198],
    [52.369553943508, 4.85309666406198],
    [52.3909553943508, 4.929309666406198],
    [52.3809553943508, 4.939309666406198],
  ]
};

describe(`Map renders correctly`, () => {
  it(`should `, () => {
    const { city, zoom, coordinates } = mockMapProps;
    const map = renderer.create(
      <Map city={city as LatLngTuple} zoom={zoom} coordinates={coordinates} activeCard={[0, 0]} />
    ).toJSON();

    expect(map).toMatchSnapshot();
  });
});
