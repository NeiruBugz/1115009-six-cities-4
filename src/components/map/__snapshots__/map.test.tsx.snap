// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map renders correctly should  1`] = `
<div
  style={
    Object {
      "height": "100%",
    }
  }
/>
`;
